torch/csrc/api/include/torch/optim/optimizer.h

```diff
@@ -26,32 +30,115 @@ class InputArchive;
+  OptimizerParamGroup(const OptimizerParamGroup& param_group) : params_(param_group.params()), options_(param_group.has_options() ? param_group.options().clone() : nullptr) {}
+  OptimizerParamGroup(std::vector<Tensor> params) : params_(std::move(params)) {}
+  OptimizerParamGroup(std::vector<Tensor> params, std::unique_ptr<OptimizerOptions> options) : params_(params), options_(std::move(options)) {}
+
+  bool has_options() const;
+  OptimizerOptions& options();
+  const OptimizerOptions& options() const;
+  void set_options(std::unique_ptr<OptimizerOptions> options);
+  std::vector<Tensor>& params();
+  const std::vector<Tensor>& params() const;
+
+ protected:
+  std::vector<Tensor> params_;
+  std::unique_ptr<OptimizerOptions> options_;
+};
+
 namespace detail {
+
 /// Base class for all optimizers, that does not yet define a `step()`
 /// mechanism. All it specifies is that optimizers must be supplied with a
 /// vector of parameters. It also defines certain methods that all optimizers
 /// shall have, such as `zero_grad`.
 class TORCH_API OptimizerBase {
  public:
+  // The copy constructor is deleted, because the user should use the
+  // `state_dict` / `load_state_dict` API to copy an optimizer instead.
+  OptimizerBase(const OptimizerBase& optimizer_base) = delete;
+  OptimizerBase(OptimizerBase&& optimizer_base) = default;
+
   /// Constructs the `Optimizer` from a vector of parameters.
   explicit OptimizerBase(std::vector<Tensor> parameters);
 
+  explicit OptimizerBase(std::vector<OptimizerParamGroup> param_groups, std::unique_ptr<OptimizerOptions> defaults) : defaults_(std::move(defaults)) {
+    for (const auto& param_group : param_groups) {
+      add_param_group(param_group);
+    }
+  }
+
+  /// Adds the given param_group to the optimizer's param_group list.
+  void add_param_group(const OptimizerParamGroup& param_group);
+
   virtual ~OptimizerBase() = default;
 
+  // TODO: when all optimizers use the new design, we can devirtualize some of the following methods
+  // such as add_parameters() / parameters() / size()
+
   /// Adds the given vector of parameters to the optimizer's parameter list.
-  void add_parameters(const std::vector<Tensor>& parameters);
+  virtual void add_parameters(const std::vector<Tensor>& parameters);
 
   /// Zeros out the gradients of all parameters.
   virtual void zero_grad();
 
   /// Provides a const reference to the parameters this optimizer holds.
-  const std::vector<Tensor>& parameters() const noexcept;
+  virtual const std::vector<Tensor>& parameters() const noexcept;
 
   /// Provides a reference to the parameters this optimizer holds.
-  std::vector<Tensor>& parameters() noexcept;
+  virtual std::vector<Tensor>& parameters() noexcept;
 
   /// Returns the number of parameters referenced by the optimizer.
-  size_t size() const noexcept;
+  virtual size_t size() const noexcept;
+
+  OptimizerOptions& defaults() noexcept;
```

Comment:
yeah I think we should also add a defaults function returning non-const reference to defaults_

Reply:
Do you mean returning a non-const reference to the `defaults_` unique_ptr? It's probably not strictly needed because we are more interested in the content of `defaults_`, which we are obtaining from the two `defaults()` functions we currently have (one const-ref and one non-const-ref)